### PR TITLE
AddPiece

### PR DESCRIPTION
**Pull request by Elestrias**
Signed-off-by: elestrias <rus8-2002@mail.ru>

### CODE REVIEW

File: core/api/rpc/json.hpp [hunk 1]
```diff
@@ -2013,6 +2014,18 @@ namespace fc::api {
...
 
+    ENCODE(MetaPieceData) {
+      Value j{rapidjson::kObjectType};
+      Set(j, "Type", v.type.toString());
+      Set(j, "Info", v.uuid);
+      return j;
+    }
+
+    DECODE(MetaPieceData) {
+      // Get(j, "Type", v.type)
```
Comment (Alexey-N-Chernyshov) [inline core/api/rpc/json.hpp:2025]: Suspicious commented line.

File: core/api/rpc/wsc.cpp [hunk 2]
```diff
@@ -107,6 +114,7 @@ namespace fc::api::rpc {
       }
       std::string_view s{static_cast<const char *>(buffer.cdata().data()),
                          buffer.cdata().size()};
+      spdlog::info("WSC READ {}", s);
```
Comment (turuslan) [inline core/api/rpc/wsc.cpp:117]: remove from pr to `master`

File: core/api/storage_miner/storage_api.cpp [hunk 3]
```diff
@@ -176,22 +178,28 @@ namespace fc::api {
...
+                                std::string sealing_mode) {
       return sector_index->storageBestAlloc(
-          allocate, sector_size, sealing_mode);
+          allocate, sector_size, (sealing_mode == "sealing"));
     };
 
     makeReturnApi(api, sector_scheduler);
 
     api->WorkerConnect =
         [=, self{api}](const std::string &address) -> outcome::result<void> {
```
Comment (turuslan) [inline core/api/storage_miner/storage_api.cpp:189]:
```suggestion
        [=](auto cb, const std::string &address) {
```

File: core/api/storage_miner/storage_api.cpp [hunk 4]
```diff
@@ -176,22 +178,28 @@ namespace fc::api {
...
+                                       // to avoid overloading
 
-      spdlog::info("Connected to a remote worker at {}", address);
+        spdlog::info("Connected to a remote worker at {}", address);
 
-      return sector_manager->addWorker(std::move(worker));
+        OUTCOME_EXCEPT(sector_manager->addWorker(std::move(worker)));
+      }).detach();
+
+      return outcome::success();
```
Comment (turuslan) [inline core/api/storage_miner/storage_api.cpp:202]:
```suggestion
        OUTCOME_CB1(sector_manager->addWorker(std::move(worker)));
        cb(outcome::success());
      }).detach();
```

File: core/api/storage_miner/storage_api.cpp [hunk 5]
```diff
@@ -176,22 +178,28 @@ namespace fc::api {
...
     api->WorkerConnect =
         [=, self{api}](const std::string &address) -> outcome::result<void> {
-      OUTCOME_TRY(maddress, libp2p::multi::Multiaddress::create(address));
-      OUTCOME_TRY(worker,
-                  RemoteWorker::connectRemoteWorker(*io, self, maddress));
+      std::thread([=]() {
+        OUTCOME_EXCEPT(
+            worker,
+            RemoteWorker::connectRemoteWorker(
+                *io, self, address));  // TODO()[FIL-] Distribute API workload
```
Comment (turuslan) [inline core/api/storage_miner/storage_api.cpp:194]:
```suggestion
      std::thread{[=] {
        OUTCOME_CB(
            auto worker,
            RemoteWorker::connectRemoteWorker(
                *io, api, address));  // TODO()[FIL-] Distribute API workload
```

File: core/sector_storage/impl/remote_worker.hpp [hunk 6]
```diff
@@ -84,7 +84,12 @@ namespace fc::sector_storage {
    private:
     explicit RemoteWorker(io_context &context);
 
+    IoThread worker_thread_;
+    IoThread httpSender_;
     WorkerApi api_;
     Client wsc_;
+    io_context &io_;
```
Comment (turuslan) [inline core/sector_storage/impl/remote_worker.hpp:91]: `io_` unused?

File: core/sector_storage/impl/remote_worker.cpp [hunk 7]
```diff
@@ -41,15 +73,90 @@ namespace fc::sector_storage {
...
+    if(!piece_data.isNullData()) {
+      PieceDataSender::send(
+          piece_data.release(),  // TODO(@Elestrias) [FIL-560] Add Null PieceData to AddPiece.
+          *(httpSender_.io),
+          host_,
+          port_,
+          "/rpc/streams/v0/push/" + meta_data.uuid,
+          new_piece_size,
+          [](const outcome::result<std::string> &res) {
+            std::cerr << res.value();
```
Comment (turuslan) [inline core/sector_storage/impl/remote_worker.cpp:156]:
```suggestion
            if (res) {
              // success, can use `res.value()`
            } else {
              // error, error code in `res.error()`
            }
```

File: core/api/rpc/json.hpp [hunk 8]
```diff
@@ -1345,13 +1366,13 @@ namespace fc::api {
     ENCODE(HealthReport) {
       Value j{rapidjson::kObjectType};
       Set(j, "Stat", v.stat);
-      Set(j, "Error", v.error);
+      //Set(j, "Error", v.error);
```
Comment (turuslan) [inline core/api/rpc/json.hpp:1369]: please remove/fix commented code

File: core/sector_storage/impl/remote_worker.hpp (hunk 6 above)
Comment (Elestrias) [reply core/sector_storage/impl/remote_worker.hpp:91]: that is for context living

File: core/api/rpc/wsc.cpp (hunk 2 above)
Comment (Elestrias) [reply core/api/rpc/wsc.cpp]: done

File: core/primitives/piece/impl/piece_data.cpp [hunk 9]
```diff
@@ -61,4 +61,18 @@ namespace fc::primitives::piece {
     other.is_null_data_ = false;
     return *this;
   }
+
+  int PieceData::release(){
+    assert(!is_null_data_);
+    int temp = fd_;
```
Comment (Markuu-s) [inline core/primitives/piece/impl/piece_data.cpp:67]: const temp

Comment (Elestrias) [reply core/primitives/piece/impl/piece_data.cpp:67]: check current commit, before review

File: core/api/rpc/json.hpp (hunk 1 above)
Comment (Elestrias) [reply core/api/rpc/json.hpp]: fixed

File: core/sector_storage/impl/remote_worker.cpp (hunk 7 above)
Comment (Elestrias) [reply core/sector_storage/impl/remote_worker.cpp]: done

**Review by turuslan — APPROVED** [review]

File: core/api/rpc/json.hpp [hunk 10]
```diff
@@ -87,8 +87,10 @@ namespace fc::api {
   using vm::actor::builtin::types::miner::WorkerKeyChange;
   using vm::actor::builtin::types::payment_channel::Merge;
   using vm::actor::builtin::types::payment_channel::
-      ModularVerificationParameter;
+  ModularVerificationParameter;
```
Comment (Alexey-N-Chernyshov) [inline core/api/rpc/json.hpp:90]: Please, revert the change.

File: core/api/rpc/json.hpp [hunk 11]
```diff
@@ -87,8 +87,10 @@ namespace fc::api {
...
   using vm::runtime::ExecutionResult;
+  using primitives::piece::PieceData;
```
Comment (Alexey-N-Chernyshov) [inline core/api/rpc/json.hpp:92]: Not used in the file.

File: core/primitives/piece/piece_data.hpp [hunk 12]
```diff
@@ -40,4 +43,35 @@ namespace fc::primitives::piece {
     bool is_null_data_;
   };
 
+  class ReaderType {
+   public:
+    enum Type : uint64_t {
+      pushStreamReader = 0,
```
Comment (Alexey-N-Chernyshov) [inline core/primitives/piece/piece_data.hpp:49]: [Enumerators should be named like constants](https://google.github.io/styleguide/cppguide.html#Enumerator_Names).
`kPushStreamReader`

File: core/remote_worker/main.cpp [hunk 13]
```diff
@@ -273,6 +273,7 @@ namespace fc {
     wrpc.emplace("/rpc/v0", api::makeRpc(*wapi));
     auto wroutes{std::make_shared<api::Routes>()};
 
+
```
Comment (Alexey-N-Chernyshov) [inline core/remote_worker/main.cpp:276]: Extra empty line.

File: core/sector_storage/impl/remote_worker.cpp [hunk 14]
```diff
@@ -3,25 +3,44 @@
...
+      explicit make_unique_enabler(io_context &context)
+          : RemoteWorker{context} {};
     };
 
     std::unique_ptr<RemoteWorker> r_worker =
         std::make_unique<make_unique_enabler>(context);
 
     r_worker->wsc_.setup(r_worker->api_);
 
+
```
Comment (Alexey-N-Chernyshov) [inline core/sector_storage/impl/remote_worker.cpp:43]: Extra empty line.

File: core/sector_storage/impl/remote_worker.cpp [hunk 15]
```diff
@@ -3,25 +3,44 @@
...
+  using common::HttpUri;
+  using primitives::jwt::kAdminPermission;
 
   outcome::result<std::shared_ptr<RemoteWorker>>
   RemoteWorker::connectRemoteWorker(io_context &context,
                                     const std::shared_ptr<CommonApi> &api,
                                     const Multiaddress &address) {
-    auto token = "stub";  // get token from common api
-
+    auto token = "stub";
```
Comment (Alexey-N-Chernyshov) [inline core/sector_storage/impl/remote_worker.cpp:32]: Is it a correct token?

File: core/sector_storage/impl/remote_worker.cpp [hunk 16]
```diff
@@ -41,15 +60,99 @@ namespace fc::sector_storage {
...
+
+    static void send(
+        int fd,
+        io_context &io,
+        const std::string &host,
+        const std::string &port,
+        const std::string &target,
+        const std::uint64_t piece_size,
+        const std::function<void(outcome::result<std::string>)> &cb) {
+      auto s{std::make_shared<PieceDataSender>(io)};
```
Comment (Alexey-N-Chernyshov) [inline core/sector_storage/impl/remote_worker.cpp:86]: Use names that describe the purpose or intent of the object. Do not worry about saving horizontal space as it is far more important to make your code immediately understandable by a new reader.
```suggestion
      auto sender{std::make_shared<PieceDataSender>(io)};
```

File: core/sector_storage/impl/remote_worker.cpp [hunk 17]
```diff
@@ -41,15 +60,99 @@ namespace fc::sector_storage {
...
+                });
+          });
+    }
+
+   private:
+    http::request<http::file_body> file_req;
+    tcp::resolver resolver;
+    beast::tcp_stream stream;
+    beast::flat_buffer buffer;
+    http::response<http::string_body> res;
```
Comment (Alexey-N-Chernyshov) [inline core/sector_storage/impl/remote_worker.cpp:124]: For private:
```suggestion
    http::request<http::file_body> file_req_;
    tcp::resolver resolver_;
    beast::tcp_stream stream_;
    beast::flat_buffer buffer_;
    http::response<http::string_body> res_;
```

File: core/api/storage_miner/storage_api.cpp [hunk 18]
```diff
@@ -72,8 +72,8 @@ namespace fc::api {
 
     api->MarketListIncompleteDeals =
         [=]() -> outcome::result<std::vector<MinerDeal>> {
-      return storage_market_provider->getLocalDeals();
-    };
+          return storage_market_provider->getLocalDeals();
```
Comment (Alexey-N-Chernyshov) [inline core/api/storage_miner/storage_api.cpp:75]: Only indents are changed, please revert the file.

File: core/sector_storage/impl/remote_worker.cpp (hunk 15 above)
Comment (Elestrias) [reply core/sector_storage/impl/remote_worker.cpp:32]: That was before me, I will change it in next PR

File: core/sector_storage/impl/remote_worker.cpp (hunk 16 above)
Comment (Elestrias) [reply core/sector_storage/impl/remote_worker.cpp]: done

File: core/sector_storage/impl/remote_worker.cpp (hunk 14 above)
Comment (Elestrias) [reply core/sector_storage/impl/remote_worker.cpp:43]: fixed

File: core/remote_worker/main.cpp (hunk 13 above)
Comment (Elestrias) [reply core/remote_worker/main.cpp:276]: fixed

File: core/api/rpc/json.hpp (hunk 11 above)
Comment (Elestrias) [reply core/api/rpc/json.hpp:92]: fixed

**Review by Alexey-N-Chernyshov — APPROVED** [review]

File: core/api/worker_api.hpp [hunk 19]
```diff
@@ -30,7 +31,13 @@ namespace fc::api {
   using sector_storage::SectorFileType;
 
   struct WorkerApi {
-    // TODO(ortyomka): [FIL-344] add AddPiece function
+    API_METHOD(AddPiece,
+               kAdminPermission,
+               CallId,
+               SectorRef,
+               gsl::span<const UnpaddedPieceSize>,
```
Comment (ortyomka) [inline core/api/worker_api.hpp]:
```suggestion
               const std::vector<const UnpaddedPieceSize> &,
```

File: core/primitives/piece/piece_data.hpp [hunk 20]
```diff
@@ -40,4 +43,29 @@ namespace fc::primitives::piece {
...
+
+    const std::string toString() const;
+  };
+
+  class MetaPieceData {
+   public:
+    MetaPieceData(std::string uuid, ReaderType::Type type)
+        : uuid(std::move(uuid)), type(ReaderType(type)){};
+    explicit MetaPieceData(): uuid("0"), type(ReaderType("null")){};
+    std::string uuid;
```
Comment (ortyomka) [inline core/primitives/piece/piece_data.hpp:67]:
```suggestion
    std::string info;
```

File: core/sector_storage/impl/remote_worker.hpp [hunk 21]
```diff
@@ -84,7 +84,11 @@ namespace fc::sector_storage {
    private:
     explicit RemoteWorker(io_context &context);
 
+    IoThread worker_thread_;
+    IoThread httpSender_;
```
Comment (ortyomka) [inline core/sector_storage/impl/remote_worker.hpp:88]: Let's get context for send from out of class
```suggestion
```

File: core/sector_storage/impl/remote_worker.hpp [hunk 22]
```diff
@@ -84,7 +84,11 @@ namespace fc::sector_storage {
    private:
     explicit RemoteWorker(io_context &context);
 
+    IoThread worker_thread_;
```
Comment (ortyomka) [inline core/sector_storage/impl/remote_worker.hpp:87]:
```suggestion
```

File: core/sector_storage/impl/remote_worker.cpp [hunk 23]
```diff
@@ -41,15 +59,100 @@ namespace fc::sector_storage {
...
+      PieceDataSender::send(
+          piece_data.release(),
+          *(httpSender_.io),
+          host_,
+          port_,
+          "/rpc/streams/v0/push/" + meta_data.uuid,
+          new_piece_size,
+          [](const outcome::result<std::string> &res) {
+            if (res.has_error()) {
+              std::cerr << res.error();
```
Comment (ortyomka) [inline core/sector_storage/impl/remote_worker.cpp:148]:
```suggestion
spdlog::error("Transfer of pieces was finished with error: {}", res.error());
```

File: core/sector_storage/impl/remote_worker.cpp [hunk 24]
```diff
@@ -41,15 +59,100 @@ namespace fc::sector_storage {
     return api_.Paths();
   }
 
-  RemoteWorker::RemoteWorker(io_context &context) : wsc_(context) {}
+  RemoteWorker::RemoteWorker(io_context &context)
+      : wsc_(*(worker_thread_.io)) {}
```
Comment (ortyomka) [inline core/sector_storage/impl/remote_worker.cpp:63]:
```suggestion
      : wsc_(context) {}
```

File: core/api/worker_api.hpp (hunk 19 above)
Comment (Elestrias) [reply core/api/worker_api.hpp]: done

File: core/sector_storage/impl/remote_worker.cpp (hunk 24 above)
Comment (Elestrias) [reply core/sector_storage/impl/remote_worker.cpp]: done

File: core/sector_storage/impl/remote_worker.cpp (hunk 23 above)
Comment (Elestrias) [reply core/sector_storage/impl/remote_worker.cpp]: done

File: core/sector_storage/impl/remote_worker.hpp (hunk 22 above)
Comment (Elestrias) [reply core/sector_storage/impl/remote_worker.hpp]: done

File: core/sector_storage/impl/remote_worker.hpp (hunk 21 above)
Comment (Elestrias) [reply core/sector_storage/impl/remote_worker.hpp]: done

File: core/primitives/piece/piece_data.hpp (hunk 20 above)
Comment (Elestrias) [reply core/primitives/piece/piece_data.hpp]: done
